langwatch/src/server/background/workers/evaluationsWorker.ts

```diff
@@ -77,6 +81,133 @@ export async function runEvaluationJob(
   });
 }
 
+/**
+ * Check if any mapping has type "thread"
+ * Single Responsibility: Detect if thread-based mappings are present
+ */
+const hasThreadMappings = (mappingState: MappingState): boolean => {
+  return Object.values(mappingState.mapping).some(
+    (mapping) => "type" in mapping && mapping.type === "thread"
+  );
+};
+
+/**
+ * Extract selected fields from traces based on thread mapping configuration
+ * Single Responsibility: Transform traces array into field values based on selectedFields
+ */
+const extractThreadFields = (
+  traces: Trace[],
+  selectedFields: string[]
+): Record<string, any>[] => {
+  return traces.map((trace) => {
+    const result: Record<string, any> = {};
+    for (const field of selectedFields) {
+      const traceMapping = TRACE_MAPPINGS[field as keyof typeof TRACE_MAPPINGS];
+      if (traceMapping) {
+        result[field] = traceMapping.mapping(trace as any, "", "", {});
+      }
+    }
+    return result;
+  });
+};
+
+/**
+ * Build thread-based data for evaluation
+ * Single Responsibility: Extract and format thread data according to thread mappings
+ */
+const buildThreadData = async (
+  projectId: string,
+  trace: Trace,
+  mappingState: MappingState,
+  protections: Protections
+): Promise<Record<string, any>> => {
+  const threadId = trace.metadata?.thread_id;
+  if (!threadId) {
+    throw new Error(
+      "Trace does not have a thread_id for thread-based evaluation"
+    );
+  }
+
+  logger.info("Fetching thread traces", {
+    threadId,
+    traceId: trace.trace_id,
+    projectId,
+  });
+
+  // Fetch all traces in the thread
+  const threadTraces = await getTracesGroupedByThreadId({
+    connConfig: { projectId },
+    threadId,
+    protections,
+    includeSpans: true,
+  });
+
+  logger.info("Thread traces fetched", {
+    threadId,
+    traceCount: threadTraces.length,
+    traceIds: threadTraces.map((t) => t.trace_id),
+  });
+
+  const result: Record<string, any> = {};
+
+  // Process each mapping
+  for (const [targetField, mappingConfig] of Object.entries(
+    mappingState.mapping
+  )) {
+    if ("type" in mappingConfig && mappingConfig.type === "thread") {
+      if (mappingConfig.source === "thread_id") {
+        result[targetField] = threadId;
+        logger.info("Mapped thread_id", {
+          targetField,
+          value: threadId,
+        });
+      } else if (mappingConfig.source === "traces") {
+        // Extract selected fields from all traces
+        const selectedFields = mappingConfig.selectedFields ?? [];
+        const extractedData = extractThreadFields(threadTraces, selectedFields);
+        result[targetField] = extractedData;
+        logger.info("Mapped thread traces", {
+          targetField,
+          selectedFields,
+          traceCount: extractedData.length,
+        });
+      }
```

Comment:
what if it's neither the first nor the second?

this can easily cause us to shoot ourselves in the foot, put a typescript guarantee here that the `else` case is completely impossible

we've done in some other parts of the codebase, look for switch statements with ` satisfies never;`

or perhaps do a hashmap where the types force you to define the mapping function of every field